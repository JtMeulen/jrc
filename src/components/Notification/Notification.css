.notification {
  font-family: "Nunito Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
  padding: 8px;
  display: flex;
  line-height: 1;
  border-radius: 8px;
  border: 1px solid lightslategray;
}

.notification__side-border {
  width: 4px;
  border-radius: 2px;
  margin-right: 8px;
}

.notification__side-border--error {
  background: red;
}

.notification__side-border--info {
  background: blue;
}

.notification__side-border--warning {
  background: orange;
}

.notification__side-border--success {
  background: green;
}
