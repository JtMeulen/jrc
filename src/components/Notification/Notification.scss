@use '../../style/lib.scss' as *;
@use '../../style/mixins.scss';

.notification {
  @include mixins.font-family();

  --jrc-color-icon: red;
  padding: 4px;
  display: flex;
  line-height: 1;
  background-color: $jrc-color-background;
  border-radius: 4px;
  border: 1px solid $jrc-color-border;
  border-left-width: 4px;

  &--hidden {
    display: none;
  }

  &--error {
    --jrc-color-icon: #{$jrc-color-error};
    border-left-color: $jrc-color-error;
  }

  &--info {
    --jrc-color-icon: #{$jrc-color-info};
    border-left-color: $jrc-color-info;
  }

  &--warning {
    --jrc-color-icon: #{$jrc-color-warning};
    border-left-color: $jrc-color-warning;
  }

  &--success {
    --jrc-color-icon: #{$jrc-color-success};
    border-left-color: $jrc-color-success;
  }

  &__icon-wrapper{
    align-self: center;
    margin: 4px;
  }

  &__text-wrapper {
    padding: 4px;
  }

  &__header {
    margin: 0px 0px 4px;
    font-size: 1em;
    font-weight: 700;
  }

  &__content {
    margin: 0px;
    font-size: 1em;
  }

  &__close-button {
    color: $jrc-color-font;
    align-self: flex-start;
    margin-left: auto;
    border: 0;
    background: transparent;
    cursor: pointer;

    &:hover, &:focus {
      color: $jrc-color-primary-hover;
    }

    &:active {
      color: $jrc-color-primary-active;
    }
  }
}
