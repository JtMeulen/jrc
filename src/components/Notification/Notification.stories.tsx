import React from "react";
import { Meta, StoryFn } from "@storybook/react";
import Notification from "./Notification";

export default {
  title: "Components/Notification",
  component: Notification
} as Meta<typeof Notification>;

const Template: StoryFn<typeof Notification> = (args) => (
  <Notification {...args} />
);

export const Default = Template.bind({});
Default.args = {
  type: "info",
  header: "I am a header",
  content: "I am content",
};

export const Dismissable = Template.bind({});
Dismissable.args = {
  type: "warning",
  header: "I am dismissable",
  content: "I am content",
  dismissable: true,
  onDismiss: () => console.log("Dismiss me"),
};

export const NoHeader = Template.bind({});
NoHeader.args = {
  type: "success",
  content: "I am content",
  dismissable: true,
  onDismiss: () => console.log("Dismiss me"),
};
