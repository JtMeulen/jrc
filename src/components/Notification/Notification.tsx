import React, { useState } from "react";
import cn from "classnames";
import { Icon } from "../index";
import "./Notification.scss";

interface NotificationProps {
  type: "error" | "info" | "warning" | "success";
  header?: string;
  content: string;
  dismissable?: boolean;
  onClose?: () => void;
}

const Notification: React.FC<NotificationProps> = ({
  type,
  header,
  content = "",
  dismissable = false,
  onClose = () => {},
}) => {
  const [hidden, setHidden] = useState(false);

  const onCloseClick = () => {
    setHidden(true);
    onClose();
  };

  return (
    <div
      className={cn("notification", `notification--${type}`, hidden && "notification--hidden")}
      data-testid="jrc-notification"
    >
      <div className="notification__icon-wrapper">
        <Icon size="sm" name={type} />
      </div>
      <div className="notification__text-wrapper">
        {header && <h2 className="notification__header">{header}</h2>}
        {content && <p className="notification__content">{content}</p>}
      </div>
      {dismissable && (
        <button
          className="notification__close-button"
          onClick={onCloseClick}
          aria-label="Close"
          data-testid="jrc-notification-close-button"
        >
          X
        </button>
      )}
    </div>
  );
};

export default Notification;
