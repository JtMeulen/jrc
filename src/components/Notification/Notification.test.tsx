import React from "react";
import { render, fireEvent,waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Notification from "./Notification";

describe("Notification", () => {
  test("renders without error", () => {
    const { container } = render(<Notification type="info" content="Notification content" />);
    expect(container).toMatchSnapshot();
  });

  // TODO: Snapshot for all props?
  test("renders with different props", async () => {
    const { container } = render(
      <Notification type="warning" content="Some content" header="Header here" dismissable />
    );
    expect(container).toMatchSnapshot();
  });

  test("hides notification and calls onDismiss handler when close button clicked", async () => {
    const handleClose = jest.fn();
    const { findByTestId } = render(
      <Notification type="warning" content="Some content" dismissable onClose={handleClose} />
    );

    const closeButtonEl = await findByTestId("jrc-notification-close-button");
    const notificationEl = await findByTestId("jrc-notification");

    fireEvent.click(closeButtonEl);
    expect(handleClose).toHaveBeenCalled();

    // CSS is not loaded correctly in the test setup, so we have to trust the hidden class for now
    await waitFor(() => {
      expect(notificationEl).toHaveClass("notification--hidden");
    });
    expect(notificationEl).toBeInTheDocument();
  });

  test("hides notification and calls onDismiss handler when enter key clicked on close button", async () => {
    const handleClose = jest.fn();
    const { findByTestId } = render(
      <Notification type="warning" content="Some content" dismissable onClose={handleClose} />
    );
    const notificationEl = await findByTestId("jrc-notification");

    await userEvent.keyboard("[Tab]");
    await userEvent.keyboard("[Enter]");
    expect(handleClose).toHaveBeenCalled();

    // CSS is not loaded correctly in the test setup, so we have to trust the hidden class for now
    await waitFor(() => {
      expect(notificationEl).toHaveClass("notification--hidden");
    });
    expect(notificationEl).toBeInTheDocument();
  });
});
