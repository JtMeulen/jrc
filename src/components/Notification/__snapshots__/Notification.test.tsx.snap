// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders with different props 1`] = `
<div>
  <div
    class="notification notification--warning"
    data-testid="jrc-notification"
  >
    <div
      class="notification__icon-wrapper"
    >
      <div
        class="icon icon--size-sm icon--warning"
      />
    </div>
    <div
      class="notification__text-wrapper"
    >
      <h2
        class="notification__header"
      >
        Header here
      </h2>
      <p
        class="notification__content"
      >
        Some content
      </p>
    </div>
    <button
      aria-label="Close"
      class="notification__close-button"
      data-testid="jrc-notification-close-button"
    >
      X
    </button>
  </div>
</div>
`;

exports[`Notification renders without error 1`] = `
<div>
  <div
    class="notification notification--info"
    data-testid="jrc-notification"
  >
    <div
      class="notification__icon-wrapper"
    >
      <div
        class="icon icon--size-sm icon--info"
      />
    </div>
    <div
      class="notification__text-wrapper"
    >
      <p
        class="notification__content"
      >
        Notification content
      </p>
    </div>
  </div>
</div>
`;
